aten/src/ATen/Error.h

```diff
@@ -25,262 +25,118 @@
-  // We always skip this frame (backtrace).
-  frames_to_skip += 1;
-
-  std::vector<void*> callstack(
-      frames_to_skip + maximum_number_of_frames, nullptr);
-  // backtrace() gives us a list of return addresses in the current call stack.
-  // NOTE: As per man (3) backtrace it can never fail
-  // (http://man7.org/linux/man-pages/man3/backtrace.3.html).
-  auto number_of_frames = ::backtrace(callstack.data(), callstack.size());
-
-  // Skip as many frames as requested. This is not efficient, but the sizes here
-  // are small and it makes the code nicer and safer.
-  for (; frames_to_skip > 0 && number_of_frames > 0;
-       --frames_to_skip, --number_of_frames) {
-    callstack.erase(callstack.begin());
-  }
-
-  // `number_of_frames` is strictly less than the current capacity of
-  // `callstack`, so this is just a pointer subtraction and makes the subsequent
-  // code safer.
-  callstack.resize(number_of_frames);
-
-  // `backtrace_symbols` takes the return addresses obtained from `backtrace()`
-  // and fetches string representations of each stack. Unfortunately it doesn't
-  // return a struct of individual pieces of information but a concatenated
-  // string, so we'll have to parse the string after. NOTE: The array returned
-  // by `backtrace_symbols` is malloc'd and must be manually freed, but not the
-  // strings inside the array.
-  std::unique_ptr<char*, std::function<void(char**)>> raw_symbols(
-      ::backtrace_symbols(callstack.data(), callstack.size()),
-      /*deleter=*/free);
-  const std::vector<std::string> symbols(
-      raw_symbols.get(), raw_symbols.get() + callstack.size());
-
-  // The backtrace string goes into here.
-  std::ostringstream stream;
-
-  for (size_t frame_number = 0; frame_number < callstack.size();
-       ++frame_number) {
-    const auto frame = parse_frame_information(symbols[frame_number]);
+inline void _str(std::stringstream& /*ss*/) {}
 
-    // frame #<number>:
-    stream << "frame #" << frame_number << ": ";
-
-    if (!frame.has_value()) {
-      // In the edge-case where we couldn't parse the frame string, we can just
-      // use it directly (it may have a different format).
-      stream << symbols[frame_number] << "\n";
-    } else {
-      // <function_name> + <offset> (<return-address> in <object-file>)
-      stream << frame->function_name << " + " << frame->offset_into_function
-             << " (" << callstack[frame_number] << " in " << frame->object_file
-             << ")\n";
-    }
-  }
-
-  return stream.str();
+template <typename T>
+inline void _str(std::stringstream& ss, const T& t) {
+  ss << t;
 }
-#endif // !defined(_WIN32)
 
-/// A tiny implementation of static `all_of`.
-template <bool...>
-struct pack;
-template <bool... values>
-struct all_of : std::is_same<pack<values..., true>, pack<true, values...>> {};
+template <typename T, typename... Args>
+inline void
+_str(std::stringstream& ss, const T& t, const Args&... args) {
```

Comment:
templates are always `inline`, no need to specify

Reply:
I don't see what the harm of specifying `inline` here is. And especially, some of the definitions I've written here are specializations of templates, and in those cases it is definitely mandatory. The rule, "Always write inline for functions written in a header" seems like a simpler rule than "Always write inline for functions written in a header, except for templated functions (but NOT template specializations; those need inline)."